### PR TITLE
Unify search logic for compilers and externals

### LINKED ISSUE #39627: Move compiler searching logic to packages
# Summary

As a core developer, I want to move the search logic for compilers from internal custom compiler classes to external support in packages so that the infrastructure will be ready when we'll move compiler from node attributes to dependencies.

# Description

At some point, during the transition to turn compiler into proper dependencies, we'll need to move the custom search logic that now populates `compilers.yaml` to packages. 

This operation can be done simultaneously with other changes, and without affecting users. The idea is to ensure that we have a package for each compiler we currently support, and add `external find` capabilities equivalent to the current search.

Then, we can add a temporary layer that translates the external information in the current `compilers.yaml` format. This layer will be removed as soon as compilers are turned into dependencies.

# Tasks

- [x] #39633
- [x] #40250
- [x] Write detection tests to replace unit-tests for compiler in core
- [x] #39780
- [x] Handle extra environment variables and extra RPATHs
- [x] Generate compiler entries from `packages.yaml`
- [ ] Use `spack external find` to search for compilers

### PR DESCRIPTION
**Pull request by alalazo**
fixes #39627

This PR is a slight step back from #44419, to refactor code in a way that doesn't affect where we write compilers.

The code used for:
```console
$ spack compiler find
```
is the same as the one used for:
```console
$ spack external find
```
Dead code in `spack.compilers` has been removed, and more can be removed later from each compiler class.

Compared to #44419, compilers are still written in `compilers.yaml` - so users shouldn't be negatively affected by this change.

### CODE REVIEW

**Comment by haampie** [comment]
`strace -f -s1024 -e execve -o out spack compiler find` shows it's not equivalent yet.

1. regexes seem to include more executables (cross compilers which are then listed with an incorrect target)
2. executables are queried repeatedly/redundantly about their version using the same flag.

This makes it 2.5x slower for me.

**Comment by alalazo** [comment]
The main issue for the slowdown you still see seems to be the overhead of loading a package (which we don't have on `develop`). This is a trace of a worker on my laptop:
Screenshot from 2024-08-16 23-49-26

The ratio for `llvm` seems to be 0.4 secs to load the pkg, 0.1 secs to detect the spec.

**Comment by alalazo** [comment]
Changes to compiler search logic to improve speed (and correctness, in case of gcc) are in #45805 and #45810

File: lib/spack/spack/compilers/__init__.py [hunk 1]
```diff
@@ -380,79 +262,75 @@ def all_compiler_specs(scope=None, init_config=True):
...
+        max_workers: number of processes used to search for compilers
     """
+    known_compilers = set(all_compilers(init_config=False))
+
     if path_hints is None:
         path_hints = get_path("PATH")
     default_paths = fs.search_paths_for_executables(*path_hints)
+    if sys.platform == "win32":
+        default_paths.extend(windows_os.WindowsOs().compiler_search_paths)
+    compiler_pkgs = spack.repo.PATH.packages_with_tags(COMPILER_TAG, full=True)
```
Comment (haampie) [inline lib/spack/spack/compilers/__init__.py:292]: import

File: lib/spack/spack/compilers/__init__.py [hunk 2]
```diff
@@ -380,79 +262,75 @@ def all_compiler_specs(scope=None, init_config=True):
...
-        item: Tuple[Optional[DetectVersionArgs], Optional[str]]
-    ) -> DetectVersionArgs:
-        value, _ = item
-        assert value is not None
-        return value
-
-    return make_compiler_list(
-        [remove_errors(detected) for detected in detected_versions if valid_version(detected)],
-        mixed_toolchain=mixed_toolchain,
+    detected_packages = spack.detection.by_path(
```
Comment (haampie) [inline lib/spack/spack/compilers/__init__.py:294]: same

**Comment by haampie** [comment]
ThreadpoolExecutor worked and works fine and is significantly faster: 3s -> 0.5s on macOS. I guess it's relatively safe although we don't control the package specific implementation... but that's unlikely to have global side effects.

**Comment by alalazo** [comment]
> ThreadpoolExecutor worked and works fine and is significantly faster: 3s -> 0.5s on macOS. I guess it's relatively safe although we don't control the package specific implementation... but that's unlikely to have global side effects.

Seems there's some work to do to use a ThreadPoolExecutor:
```console
$ spack external find -t compiler
==> Warning: error detecting "llvm" from prefix /bin [trying to set variants "clang, lld, or lldb" in package "llvm", but the package has no such variants [happened when validating 'llvm@10.0.0+clang~lld~lldb']]
==> Warning: error detecting "llvm" from prefix /usr/bin [trying to set variants "clang, lld, or lldb" in package "llvm", but the package has no such variants [happened when validating 'llvm@12.0.0+clang+lld~lldb']]
==> No new external packages detected
```
vs:
```console
$ spack external find -t compiler
==> The following specs have been detected on this system and added to /home/culpo/.spack/packages.yaml
gcc@8.4.0  gcc@9.4.0  gcc@10.5.0  llvm@10.0.0  llvm@11.0.0  llvm@12.0.0
```

**Review by haampie — APPROVED** [review]

**Pull request merged**